[RFC] main: make --embed wait for first request so embedding UI can display startup messages

bfredl opened the pull request:
This allows to properly handle first buffer swap message, as well as
errors in startup files and --cmd commands in an external UI starting nvim as a child, just by using `--embed-ui` option instead of `--embed`

Note: this does nothing that a `rpcrequest()` in a `--cmd` and a corresponding handler in the UI client cannot do. But I think default matters, making a UI that has non-racy startup should be _simple_ out of the box, rather than the UI needing to know details of the nvim startup process and event handling model even if it only ever wants to use `nvim_ui_attach` and `nvim_input`.

bfredl commented:
Also, later on `nvim_ui_attach` should be made non-deferred, but it might require some refactors, for instance in a wait-return state it cannot redraw the screen, it should rather just retransmit the screen (or even just fake the message if the new client wants to be narrower than the current message displayed before the wait-return prompt). That will now however only affect connecting to running instance, this PR handles the start-up case.

jamessan reviewed (approved):
LGTM other than the lint warning

oni-link commented on src/nvim/main.c:
```diff
@@ -819,6 +829,14 @@ static void command_line_scan(mparm_T *parmp)
             mch_exit(0);
           } else if (STRICMP(argv[0] + argv_idx, "headless") == 0) {
             headless_mode = true;
+          } else if (STRICMP(argv[0] + argv_idx, "embed-ui") == 0) {
```
Will this option (later) be added to `nvim --help` and the internal help?

bfredl replied:
It will be added now.

bfredl commented:
Yes, I will add some docs and test also :)

bfredl commented:
Messages sometimes get truncated, so this needs some more work.

bfredl commented:
After some discussion with @justinmk I think we could use the following rule to avoid adding more an option: 

with `--embed` nvim always waits for _any_ request on channel 1 before continuing with startup, this will break some clients but not all non-ui `--embed`-ers as using `--embed-ui` as `--embed` would. The old behavior can still be recovered with `nvim --headless --cmd "call stdioopen({'rpc': v:true})"`

bfredl committed:
startup: wait for embedder before executing startup commands and files

Give embeders a chance to set up nvim, by processing a request before
startup. This allows an external UI to show messages and prompts from
--cmd and buffer loading (e.g. swap files)

bfredl commented:
Update: now handles messages more consistently with TUI (never show more than one "press RETURN" prompt, unless the screen is really full or something). Note: unlike early TUI early messages can here be affected by colors loaded by vimrc. Which means if one message was displayed before `:color` and one after, the first will appear with builtin colors and the second with user colors. I think this is better than (1) wiping away the earliest message before it can be seen or (2) showing two "press RETURN" prompts just to not mix messages with different color schemes. Anyway this situation should be pretty rare. (swap messages, which should be the most common, will be shown with the user colors, just like TUI).

justinmk commented on test/functional/helpers.lua:
```diff
@@ -362,6 +362,7 @@ local function clear(...)
     table.insert(args, arg)
   end
   set_session(spawn(args, nil, env))
+  session:request('nvim_eval', "0")
```
How about a comment:

    -- Dummy request so that --embed starts the UI.

bfredl commented on src/nvim/generators/gen_api_dispatch.lua:
```diff
@@ -128,6 +128,7 @@ for i,f in ipairs(shallowcopy(functions)) do
     end
     newf.impl_name = f.name
     newf.remote_only = true
+    newf.async = f.async
```
I must have forgotten to remove this line, `shallowcopy(f)` already handles this...

bfredl commented:
Updated with a note in   `starting.txt`, I will add a test or two also.

bfredl commented:
Added tests.

justinmk commented on src/nvim/msgpack_rpc/channel.c:
```diff
@@ -40,6 +40,8 @@
 static PMap(cstr_t) *event_strings = NULL;
 static msgpack_sbuffer out_buffer;
 
+static bool seen_stdio_request = false;
```
`got_stdio_request` is more common convention.
